src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -364,13 +389,26 @@ public override bool MoveNext()
                 {
                     case 1:
                         _enumerator = _source.GetEnumerator();
-                        state = 2;
-                        goto case 2;
+                        if (_predicate == null)
```

Comment:
This is never going to change for the object's lifetime, so there's no need to check it more than once per enumeration.